### PR TITLE
Adds JSON schema support for skaffold.yaml

### PR DESCRIPTION
**Pull request by etanshaul**
Embeds the schemas into the plugin for this PR. Later PR will load dynamically.

- [x] Get rid of unnecessary `skaffold-editing` module and merge it with `skaffold`
- [x] POC json schema validation with embedded schema
- [x] Automatically select schema version based on skaffold version
- [x] Testing
- [ ] Expose a UI to select a skaffold version? (Not sure if we need this, but not for this PR)

### CODE REVIEW

**Comment by etanshaul** [comment]
image
image

**Comment by etanshaul** [comment]
@ivanporty fyi. working on test coverage

**Comment by etanshaul** [comment]
@ivanporty ready for review

File: kubernetes/skaffold/src/main/resources/schemas/v1alpha1.json [hunk 1]
```diff
@@ -0,0 +1,216 @@
+{
```
Comment (ivanporty) [inline kubernetes/skaffold/src/main/resources/schemas/v1alpha1.json]: How do we maintain these files in the future? Where are they copied from?

File: kubernetes/skaffold/src/test/kotlin/com/google/container/tools/skaffold/editing/SkaffoldSchemaFileProviderTest.kt [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+    }
+
+    @Test
+    fun `isAvailable returns false when not matching skaffold schema`() {
+        every { skaffoldFileService.getSkaffoldVersion(any()) } answers { "v1alpha1" }
+        every { skaffoldFileService.isSkaffoldFile(any()) } answers { true }
+
+        assertThat(skaffoldSchemaFileProvider.isAvailable(virtualFile)).isFalse()
+    }
+}
```
Comment (ivanporty) [inline kubernetes/skaffold/src/test/kotlin/com/google/container/tools/skaffold/editing/SkaffoldSchemaFileProviderTest.kt]: EOL

File: kubernetes/skaffold/src/main/kotlin/com/google/container/tools/skaffold/SkaffoldFileService.kt [hunk 3]
```diff
@@ -61,6 +81,11 @@ class SkaffoldFileService {
      */
     fun findSkaffoldFiles(project: Project): List<VirtualFile> {
         return FileTypeIndex.getFiles(YAMLFileType.YML, GlobalSearchScope.allScope(project))
-            .filter { isSkaffoldFile(it) }
+                .filter { isSkaffoldFile(it) }
     }
+
+    private fun isValidSkaffoldFileType(file: VirtualFile): Boolean =
```
Comment (ivanporty) [inline kubernetes/skaffold/src/main/kotlin/com/google/container/tools/skaffold/SkaffoldFileService.kt]: This was done as a preliminary check for YAML files, there's nothing particular about Skaffold, how about `isValidYAMLFileType`?

File: kubernetes/skaffold/src/main/kotlin/com/google/container/tools/skaffold/editing/SkaffoldSchemaFileProvider.kt [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+    : JsonSchemaFileProvider {
+
+    override fun getName(): String = SKAFFOLD_SCHEMA_NAME
+
+    /**
+     * Returns true if the currently open file is a valid Skaffold file, and the Skaffold version
+     * associated with this schema provider matches the Skaffold version in the currently open file.
+     */
+    override fun isAvailable(file: VirtualFile): Boolean {
+        val skaffoldVersion: String? = SkaffoldFileService.instance.getSkaffoldVersion(file)
```
Comment (ivanporty) [inline kubernetes/skaffold/src/main/kotlin/com/google/container/tools/skaffold/editing/SkaffoldSchemaFileProvider.kt]: Probably makes sense to inline this into return expression, so we can short-cut the check if file is not Skaffold file at all

File: kubernetes/skaffold/src/main/kotlin/com/google/container/tools/skaffold/editing/SkaffoldSchemaFileProvider.kt [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+
+        return SkaffoldFileService.instance.isSkaffoldFile(file) &&
+                skaffoldVersion == skaffoldSchemaVersion
+    }
+
+    override fun getSchemaFile(): VirtualFile? {
+        return JsonSchemaProviderFactory
+                .getResourceFile(this::class.java, "/schemas/$skaffoldSchemaVersion.json")
```
Comment (ivanporty) [inline kubernetes/skaffold/src/main/kotlin/com/google/container/tools/skaffold/editing/SkaffoldSchemaFileProvider.kt]: Convenient method! Is this the one used for other schemas too?

File: kubernetes/skaffold/src/main/resources/schemas/v1alpha1.json (hunk 1 above)
Comment (etanshaul) [reply kubernetes/skaffold/src/main/resources/schemas/v1alpha1.json]: they are currently created in the skaffold repo (on each release). This PR is the first pass at quickly getting the plumbing in place and getting support for schema editing. As is, we'd need to copy in new schemas into this repo to get support for new Skaffold versions.

The goal next is to dynamically load them (ideally from a GCS bucket) so that we don't have to maintain anything.

File: kubernetes/skaffold/src/main/kotlin/com/google/container/tools/skaffold/SkaffoldFileService.kt (hunk 3 above)
Comment (etanshaul) [reply kubernetes/skaffold/src/main/kotlin/com/google/container/tools/skaffold/SkaffoldFileService.kt]: oh yeah.. good point. 

I wonder now though if `getSkaffoldVersion` should also first check if its a valid skaffold file too..

File: kubernetes/skaffold/src/main/kotlin/com/google/container/tools/skaffold/editing/SkaffoldSchemaFileProvider.kt (hunk 5 above)
Comment (etanshaul) [reply kubernetes/skaffold/src/main/kotlin/com/google/container/tools/skaffold/editing/SkaffoldSchemaFileProvider.kt]: currently this is only used to load the Skaffold schemas. I'm not sure if that is what you were thinking of though?

File: kubernetes/skaffold/src/main/kotlin/com/google/container/tools/skaffold/editing/SkaffoldSchemaFileProvider.kt (hunk 4 above)
Comment (etanshaul) [reply kubernetes/skaffold/src/main/kotlin/com/google/container/tools/skaffold/editing/SkaffoldSchemaFileProvider.kt]: great catch. Updated to do the short-circuiting

File: kubernetes/skaffold/src/main/kotlin/com/google/container/tools/skaffold/editing/SkaffoldSchemaFileProvider.kt (hunk 5 above)
Comment (ivanporty) [reply kubernetes/skaffold/src/main/kotlin/com/google/container/tools/skaffold/editing/SkaffoldSchemaFileProvider.kt]: Yeah, I thought it's convenient to just convert class resource into a virtual file, they need to make it more broad eventually. Just interested if Jetbrains code uses it as well for schema loading.

File: kubernetes/skaffold/src/main/resources/schemas/v1alpha1.json (hunk 1 above)
Comment (ivanporty) [reply kubernetes/skaffold/src/main/resources/schemas/v1alpha1.json]: Yeah, that'd be better, kind of maintenance burden to do this here in the code, huge set of data checked in the code base.

File: kubernetes/skaffold/src/main/kotlin/com/google/container/tools/skaffold/SkaffoldFileService.kt (hunk 3 above)
Comment (etanshaul) [reply kubernetes/skaffold/src/main/kotlin/com/google/container/tools/skaffold/SkaffoldFileService.kt]: refactoring now

File: kubernetes/skaffold/src/test/kotlin/com/google/container/tools/skaffold/editing/SkaffoldSchemaFileProviderTest.kt [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+    fun `isAvailable returns false when invalid skaffold file`() {
+        every { skaffoldFileService.getSkaffoldVersion(any()) } answers { SKAFFOLD_VERSION }
+        every { skaffoldFileService.isSkaffoldFile(any()) } answers { false }
+
+        assertThat(skaffoldSchemaFileProvider.isAvailable(virtualFile)).isFalse()
+    }
+
+    @Test
+    fun `isAvailable returns false when not matching skaffold schema`() {
+        every { skaffoldFileService.getSkaffoldVersion(any()) } answers { "v1alpha1" }
```
Comment (ivanporty) [inline kubernetes/skaffold/src/test/kotlin/com/google/container/tools/skaffold/editing/SkaffoldSchemaFileProviderTest.kt]: Should we add another test when `getSkaffoldVersion` returns null - even if says this is a Skaffold file - like some incompatible changes?

File: kubernetes/skaffold/src/main/resources/schemas/v1alpha1.json (hunk 1 above)
Comment (etanshaul) [reply kubernetes/skaffold/src/main/resources/schemas/v1alpha1.json]: agreed. are you ok with phasing it though (as in deferring the enhancement to another PR)? They haven't automated pushing these to a GCS bucket yet.

File: kubernetes/skaffold/src/main/kotlin/com/google/container/tools/skaffold/editing/SkaffoldSchemaFileProvider.kt (hunk 5 above)
Comment (etanshaul) [reply kubernetes/skaffold/src/main/kotlin/com/google/container/tools/skaffold/editing/SkaffoldSchemaFileProvider.kt]: yeah this is how the platform does this as well

File: kubernetes/skaffold/src/main/resources/schemas/v1alpha1.json (hunk 1 above)
Comment (ivanporty) [reply kubernetes/skaffold/src/main/resources/schemas/v1alpha1.json]: Sure, let's keep it here now - long waited schema editing is welcome! Good step for people to easier integrate existing projects with the plugin Skaffold features.

File: kubernetes/skaffold/src/main/kotlin/com/google/container/tools/skaffold/SkaffoldFileService.kt (hunk 3 above)
Comment (etanshaul) [reply kubernetes/skaffold/src/main/kotlin/com/google/container/tools/skaffold/SkaffoldFileService.kt]: I refactored it to have `getSkaffoldVersion` also check that its a valid skaffold file, while avoiding any duplication of code. 

I'm not sure if I made it worse though..ptal.

File: kubernetes/skaffold/src/test/kotlin/com/google/container/tools/skaffold/editing/SkaffoldSchemaFileProviderTest.kt (hunk 6 above)
Comment (etanshaul) [reply kubernetes/skaffold/src/test/kotlin/com/google/container/tools/skaffold/editing/SkaffoldSchemaFileProviderTest.kt]: added

File: kubernetes/skaffold/src/main/kotlin/com/google/container/tools/skaffold/SkaffoldFileService.kt [hunk 7]
```diff
@@ -38,19 +38,37 @@ class SkaffoldFileService {
...
+            false
+        }
+    }
+
+    /**
+     * Returns the version of Skaffold as defined in skaffold.yaml, or null if it can't be read.
+     *
+     * The version is represented as skaffold/[version] in the yaml.
+     */
+    fun getSkaffoldVersion(file: VirtualFile): String? {
```
Comment (ivanporty) [inline kubernetes/skaffold/src/main/kotlin/com/google/container/tools/skaffold/SkaffoldFileService.kt]: Looking at this method and its using YAML data, how about actually moving it into `SkaffoldYamlConfiguration` itself?

Comment (ivanporty) [reply kubernetes/skaffold/src/main/kotlin/com/google/container/tools/skaffold/SkaffoldFileService.kt]: At least the version extraction part

Comment (etanshaul) [reply kubernetes/skaffold/src/main/kotlin/com/google/container/tools/skaffold/SkaffoldFileService.kt]: agreed. ill do that

Comment (etanshaul) [reply kubernetes/skaffold/src/main/kotlin/com/google/container/tools/skaffold/SkaffoldFileService.kt]: done

File: kubernetes/skaffold/src/main/kotlin/com/google/container/tools/skaffold/SkaffoldFileService.kt [hunk 8]
```diff
@@ -38,19 +38,35 @@ class SkaffoldFileService {
...
+        return try {
+            if (isValidYamlFile(file)) {
+                val skaffoldYaml = SkaffoldYamlConfiguration(file)
+
+                if (isValidSkaffoldFile(skaffoldYaml)) {
+                    skaffoldYaml.skaffoldVersion
                 }
             }
+
+            null
```
Comment (ivanporty) [inline kubernetes/skaffold/src/main/kotlin/com/google/container/tools/skaffold/SkaffoldFileService.kt]: no need for this null anymore?

Comment (etanshaul) [reply kubernetes/skaffold/src/main/kotlin/com/google/container/tools/skaffold/SkaffoldFileService.kt]: I think we still need it no? for example if `isValidYamlFile` returns false. Removing it causes it not to compile too since it leaves a dangling condition.

Comment (ivanporty) [reply kubernetes/skaffold/src/main/kotlin/com/google/container/tools/skaffold/SkaffoldFileService.kt]: ah I see. good to go then!
